### PR TITLE
T2D-141 Added better error report

### PR DESCRIPTION
**Pull request by jorizci**
-Added new exception to report when there missing unsaved exceptions.
-Added new exception handler for basic rest service.
-Added auto configuration for the exception handler using annotations.
-Added Tests
-Moved exceptions to core

### CODE REVIEW

File: src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/BasicRestControllerAdvice.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+public class BasicRestControllerAdvice {
+
+    private static final Logger logger = LoggerFactory.getLogger(BasicRestControllerAdvice.class);
+
+    @ExceptionHandler(value = {AccessionIsNotPending.class, AccessionCouldNotBeGeneratedException.class,
+            MissingUnsavedAccessions.class})
+    public ResponseEntity<ErrorMessage> handleInternalServerErrors(Exception ex) {
+        logger.error(ex.getMessage(), ex);
+        return new ResponseEntity<>(
+                new ErrorMessage(HttpStatus.INTERNAL_SERVER_ERROR, ex.getMessage()),
```
[the remark this answers is not in the document]
Comment (jorizci) [reply src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/BasicRestControllerAdvice.java]: If you want a simple message without structure, no. If you want a specific rest message structure, then you need one. I opted for the second approach.

File: src/test/java/uk/ac/ebi/ampt2d/test/rest/MockTestAccessioningService.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        for (BasicRestModel message : messages) {
+            generateAccession(message);
+        }
+        return messages.stream().collect(Collectors.toMap(
+                o -> valueToAccession.get(o.getValue()),
+                o -> o
+        ));
+    }
+
+    private synchronized void generateAccession(BasicRestModel model) throws AccessionCouldNotBeGeneratedException {
```
[the remark this answers is not in the document]
Comment (jorizci) [reply src/test/java/uk/ac/ebi/ampt2d/test/rest/MockTestAccessioningService.java]: The AccessionCouldNotBeGenerated comes from the accession generator, the other come from the accessioning service. When I designed the first one made it as a checked exception, as I thought it was a real posibility and all the interfaces followed this. Accession is not pending was created as runtime as it is not expected unless a coding bug. When I added the new one, as it is an edge case that should not happen I followed the same convention.

File: src/test/java/uk/ac/ebi/ampt2d/test/rest/BasicRestModel.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ */
+package uk.ac.ebi.ampt2d.test.rest;
+
+public class BasicRestModel {
+
+    private String value;
+
+    BasicRestModel() {
+    }
```
[the remark this answers is not in the document]
Comment (jorizci) [reply src/test/java/uk/ac/ebi/ampt2d/test/rest/BasicRestModel.java]: Spring does, that is why it is package protected.

File: src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/BasicRestControllerAdvice.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jorizci) [reply src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/BasicRestControllerAdvice.java]: The error shown in that post is the default message for exceptions generated by anything that Spring controls, mainly validation. Any exception that gets away of that ends captured at spring FrameworkServlet that throws a plain exception, makes a big 500 error with the full exception stack, no nice message.

Now, the class that it mentions it is not used in Spring as a structure, but it is part of the whole spring boot magic dust configuration. See at: 'https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/web/servlet/error/DefaultErrorAttributes.html' That is why nobody uses it on any tutorial.

[the remark this answers is not in the document]
Comment (jorizci) [reply src/main/java/uk/ac/ebi/ampt2d/commons/accession/rest/BasicRestControllerAdvice.java]: After playing for a bit with that class, it's better to do not use it, if you want I can use the same structure, but trying to use it is way far complex.
